sync/server.md

```diff
@@ -0,0 +1,404 @@
+
+* Open the Data Sync Admin UI
+* Define the Data Schema
+* Configure Data Source(s)
+* Configure Resolvers for any Queries or Mutations
+* Update/modify the Data Schema as needed, continuously saving and trying out the Schema
+
+The last part of the flow is very important.
+This implies a fast feedback loop when developing your Data Schema.
+To ensure the feedback loop is as fast as possible, the architecture of sync will be slightly different that the proof of concept architecture.
+Both the backend for the Admin UI, and the GraphQL API that Apps talk to will be part of the 1 Data Sync server.
+This server will continously run, without any restart needed when the Data Schema changes.
+To facilitate this, the mounting of the graphql express middleware will need to be done in such a way that it always rebuilds the schema based on the latests available. In reality, the parsed schema should be cached, and only reloaded/reparsed if it has changed. Here is a sample graphql endpoint mounting in express:
+
+```
+const tracing = true
+app.use('/graphql', bodyParser.json(), function (req, res, next) {
+  const schema = require('./lib/schemaParser').parseFromFile(SCHEMA_FILE, DATA_SOURCES_FILE, RESOLVER_MAPPINGS_FILE)
+  const graphql = graphqlExpress({ schema, tracing })
+  graphql(req, res, next)
+})
+```
+
+As the server needs to store resources, such as the Data Schema, Resolver Mappings and Webhooks, a data store is required.
+The Data Sync server is intended to run in a Kubernetes environment such as Openshift.
+This allows the Data Sync server to use a ServiceAccount token to save these resources in 1 or more Secrets.
+This will avoid the need for a potentially heavyweight database, at least until something different is required.
+
+A pubsub service will be required for managing the event driven nature of Subscriptions when the Data Sync server is scaled beyond 1 Replica.
+A reasonable candidate for this is Redis and its set of [pubsub commands](https://redis.io/topics/pubsub).
+Any Clients interested in a particular event/subscription will connect via websocket to the server.
+Each server replica will connect to the same redis instance and subscribe to relevant 'Subscription' events, and notify connected clients of those events.
+
+The Admin UI will be written in React, and use Patternfly for the look & feel.
+See https://github.com/patternfly/patternfly-react & https://github.com/patternfly/patternfly-react-demo-app for more info on the technology.
+See [Sync UI](./ui.md) for more info on the Admin UI Architecture
+
+## PostgreSQL Connector
+
+### Data Source
+
+The defintion for a PostgreSQL Data Source will have the following parameters:
+
+* Database Server Host
+* Database Server Port
+* Database Name
+* Database Username
+* Database Password
+
+Table names will not be configured as part of the Data Source for Postgres.
+Instead, the table will be determined by the SQL query defined in the Resolver Request Mapping.
```

Comment:
Agreed, that's a great idea.
We have this in POC schema already:
```
type Note {
  id: String
  title: String
  content: String
  timestamp: Float
}
```
We would just make the id field a GraphQL id.

Reply:
Quick question about this though - How do we ensure the table is created. Let's say the user does the following:

* defines a schema
* defines a postgres data source (no table config)
* defines the a request mapping to `SELECT * from foo`

How do we ensure that the `foo` table gets created in the first place?

Reply:
What I understand from the comment https://github.com/aerogear/proposals/pull/40/files#r198175673 , we will have a special endpoint which the schema is posted. I think that should be the step to create tables.

BTW, we also need to alter the tables if types are changed in the schema. ...which should be easy.

Reply:
@aliok data sync is not concerned with the data source, beyond configuring how to connect to it.
Data sync will have supported data sources e.g. postgres & HTTP,
but only in the context of resolver mappings.
Data Sync will not ship with a db backend for storing app/user data.

Reply:
Ok, then what's the answer to Dara's question @david-martin ?

Reply:
> Quick question about this though - How do we ensure the table is created. Let's say the user does the following:
> 
> defines a schema
> defines a postgres data source (no table config)
> defines the a request mapping to SELECT * from foo
> How do we ensure that the foo table gets created in the first place?

We don't.
It is up to the developer to have their data source/database in the right condition for resolvers to work.
If the host is wrong, port is wrong, credentials are wrong, table is missing, fields are missing, data sync will report back an error, which the developer will have to fix.